src/gpu/multiexp.rs

```diff
@@ -135,22 +130,23 @@ impl<E> SingleMultiexpKernel<E> where E: Engine {
 }
 
 // A struct that containts several multiexp kernels for different devices
-pub struct MultiexpKernel<E>(Vec<SingleMultiexpKernel<E>>) where E: Engine;
+pub struct MultiexpKernel<E> where E: Engine {
+    kernels: Vec<SingleMultiexpKernel<E>>,
+    chunk_size: usize
+}
 
 impl<E> MultiexpKernel<E> where E: Engine {
 
-    pub fn create() -> GPUResult<MultiexpKernel<E>> {
-        let devices = &utils::GPU_NVIDIA_DEVICES;
-        if devices.len() == 0 { return Err(GPUError {msg: "No working GPUs found!".to_string()} ); }
-        let mut kernels = Vec::new();
-        for dev in devices.iter().map(|device| { SingleMultiexpKernel::<E>::create(*device, CHUNK_SIZE as u32) }) {
-            kernels.push(dev?);
-        }
+    pub fn create(chunk_size: usize) -> GPUResult<MultiexpKernel<E>> {
+        if BLS12_KERNELS.len() == 0 { return Err(GPUError {msg: "No working GPUs found!".to_string()} ); }
+        let kernels : Vec<_> = BLS12_KERNELS.iter().map(|pq| {
+            SingleMultiexpKernel::<E>::create(pq.clone(), chunk_size as u32)
+        }).filter(|res| res.is_ok()).map(|res| res.unwrap()).collect();
```

Comment:
if you have an iterator of results you can collect into `Result<Vec<_>, _>`, and only unwrap/handle the error at the end

Reply:
hmm I want to filter the results that don't have any errors. It is not necessary to initialize all devices. Only initializing those that work and using them is a better strategy I think. Or maybe you mean something else?

Reply:
sorry I missread the code, this should do the same thing

```rust
let kernels = BLS12_KERNELS.iter().filter_map(|pq| {
  SingleMultiexpKernel::<E>::create(pq.clone(), chunk_size as u32).ok()
}).collect::<Vec<_>>().unwrap();
```

Reply:
this fails unfortunately, ```no method named `unwrap` found for type `std::vec::Vec<std::option::Option<gpu::multiexp::SingleMultiexpKernel<E>>>```

Reply:
nvm, just leave it as it is,